clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestFuture.java

```diff
@@ -128,26 +126,34 @@ public void raise(Errors error) {
         raise(error.exception());
     }
 
-    private void fireSuccess() {
-        for (RequestFutureListener<T> listener : listeners)
+    private void fireSuccess(T value) {
+        while (true) {
+            RequestFutureListener<T> listener = listeners.poll();
+            if (listener == null)
+                break;
             listener.onSuccess(value);
+        }
```

Comment:
I am OK with leaving it as is. Looking more carefully, we are now removing the listener on firing success. Is the idea to avoid memory leaks if the future is held in memory?